### PR TITLE
Allow configuring of SourceKit-LSP’s options using configuration files

### PR DESCRIPTION
**Pull request by ahoppen**
The idea here is to unify the different ways in which we can currently set options on SourceKit-LSP in a scalable way: Environment variables, command line arguments to `sourcekit-lsp` and initialization options.

The idea is that a user can define a `~/.sourcekit-lsp/.sourcekit-lsp` file (we store logs in `~/.sourcekit-lsp/logs` on non-Darwin platforms), which will be used as the default configuration for all SourceKit-LSP instances. They can also place a `.sourcekit-lsp` file in the root of a workspace to configure SourceKit-LSP for that project specifically, eg. setting arguments that need to be passed to `swift build` for that project and which thus also need to be set on SourceKit-LSP.

For compatibility reasons, I’m mapping the existing command line options into the new options structure for now. I hope to delete the command line arguments in the future and solely rely on `.sourcekit-lsp` configuration files.

Environment variable will be migrated to `.sourcekit-lsp` in a follow-up commit.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Documentation/Configuration File.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# `.sourcekit-lsp` Configuration File
+
+`.sourcekit-lsp` configuration files can be used to modify the behavior of SourceKit-LSP in various ways. The following locations are checked:
+- `~/.sourcekit-lsp/.sourcekit-lsp` (ie. a `.sourcekit-lsp` file in the `~/.sourcekit-lsp` directory)
+- A `.sourcekit-lsp` file in a workspace’s root
```
Comment (bnbarham) [inline Documentation/Configuration File.md:5]: What about `.sourcekit-lsp/config` and require the directory in the workspace root as well? Seems plausible we may add more later. And IMO ideally we'd check `Library/org.swift.sourcekit-lsp/config` for macOS -> `$XDG_CONFIG_HOME/sourcekit-lsp/config` everywhere (though not sure what the standard is for Windows) -> `.sourcekit-lsp/config`.

I do wonder if we need to keep the server options still as well - consider experimental features that may only be supported by specific editors + a user that jumps between editors. Though arguably that case might not work even if you *could* configure them differently. This also makes it more difficult for editors to add support for specific options.

Comment (ahoppen) [reply Documentation/Configuration File.md]: Good ideas. I modified the search locations slightly to 

- `~/.sourcekit-lsp/config.json`
- On macOS: `~/Library/Application Support/org.swift.sourcekit-lsp/config.json` from the various `Library` folders on the system
- If the `XDG_CONFIG_HOME` environment variable is set: `$XDG_CONFIG_HOME/org.swift.sourcekit-lsp/config.json`
- A `.sourcekit-lsp/config.json` file in a workspace’s root

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

File: Sources/SourceKitLSP/SourceKitLSPServer.swift [hunk 2]
```diff
@@ -989,27 +958,9 @@ extension SourceKitLSPServer {
...
-          break
-        case .some(.null):
-          self.options.completionOptions.maxResults = nil
-        case .some(.int(let maxResults)):
-          self.options.completionOptions.maxResults = maxResults
-        case .some(let invalid):
-          logger.error("Expected null or int for 'maxResults'; got \(String(reflecting: invalid))")
-        }
-      }
-    }
```
Comment (hamishknight) [inline Sources/SourceKitLSP/SourceKitLSPServer.swift:1008]: I guess this isn't supported anymore?

File: Sources/LanguageServerProtocol/Requests/CompletionRequest.swift [hunk 3]
```diff
@@ -36,18 +36,14 @@ public struct CompletionRequest: TextDocumentRequest, Hashable {
 
   public var context: CompletionContext?
 
-  public var sourcekitlspOptions: SKCompletionOptions?
-
   public init(
     textDocument: TextDocumentIdentifier,
     position: Position,
-    context: CompletionContext? = nil,
-    sourcekitlspOptions: SKCompletionOptions? = nil
```
Comment (hamishknight) [inline Sources/LanguageServerProtocol/Requests/CompletionRequest.swift:45]: There's a doc comment for this parameter above, I guess it should be removed?

**Review by hamishknight — APPROVED** [review]

File: Sources/SKCore/SourceKitLSPOptions.swift [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        maxCoresPercentageToUseForBackgroundIndexing: override?.maxCoresPercentageToUseForBackgroundIndexing
+          ?? base.maxCoresPercentageToUseForBackgroundIndexing,
+        updateIndexStoreTimeout: override?.updateIndexStoreTimeout ?? base.updateIndexStoreTimeout
+      )
+    }
+  }
+
+  public var swiftPM: SwiftPMOptions?
+  public var compilationDatabase: CompilationDatabaseOptions?
+  public var fallbackBuildSystem: FallbackBuildSystemOptions?
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SKCore/SourceKitLSPOptions.swift:142]: I tried that as well but then JSON decoding failed if the keys didn’t exist

File: Sources/SKCore/SourceKitLSPOptions.swift [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    /// Build configuration (debug|release).
+    public var configuration: BuildConfiguration?
+
+    /// Build artifacts directory path. If nil, the build system may choose a default value.
+    public var scratchPath: String?
+
+    public var cCompilerFlags: [String]?
+    public var cxxCompilerFlags: [String]?
+    public var swiftCompilerFlags: [String]?
+    public var linkerFlags: [String]?
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SKCore/SourceKitLSPOptions.swift:34]: Yes, there’s a difference between `nil` and empty here:

Say you have you global configuration file with `cCompilerFlags: ["-DFLAG"]`. And then there’s a workspace-level configuration file with `cCompilerFlags: []`. In that case `cCompilerFlags` for the workspace is empty. But if the workspace-level configuration file didn’t contain the `cCompilerFlags` key, then we’d inherit the global `cCompilerFlags` of `-DFLAG`.

File: Sources/SourceKitLSP/SourceKitLSPServer.swift (hunk 2 above)
Comment (ahoppen) [reply Sources/SourceKitLSP/SourceKitLSPServer.swift:1008]: Yes, I’m not aware of any uses for it and I don’t see it as useful. We can always add it again if we should need it, at which point we probably wouldn’t make it in initialization option anymore either.

File: Sources/SKCore/SourceKitLSPOptions.swift [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    self.workDoneProgressDebounce = workDoneProgressDebounce
+  }
+
+  public init?(path: URL?) {
+    guard let path, let contents = try? String(contentsOf: path, encoding: .utf8) else {
+      return nil
+    }
+    guard
+      let decoded = orLog(
+        "Parsing .sourcekit-lsp",
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SKCore/SourceKitLSPOptions.swift:209]: Thanks for spotting. Forgot to rename.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

File: Sources/SKCore/SourceKitLSPOptions.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (hamishknight) [reply Sources/SKCore/SourceKitLSPOptions.swift:142]: Ah yeah, I forgot Decodable doesn't utilize property initializers as default values :(

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Review by bnbarham — APPROVED** [review]
IMO the XDG_CONFIG_HOME should be just sourcekit-lsp rather than reverse-domain. I'd also specify the configuration in LSP and then allow taking it in initializationOptions. But feel free to merge and fix in a follow up 👍

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Pull request merged**